pkg/runtime/bulk_subscriber.go

```diff
@@ -0,0 +1,706 @@
+	if entryID == "" {
+		log.Warn("Invalid blank entry id received while processing bulk pub/sub event, won't be able to process it")
+		return errors.New("Blank entryID supplied - won't be able to process it")
+	}
+	return nil
+}
+
+func populateBulkSubcribedMessage(message *pubsub.BulkMessageEntry, event interface{},
+	routePathBulkMessageMap *map[string]pubsubBulkSubscribedMessage,
+	rPath string, i int, msg *pubsub.BulkMessage, isCloudEvent bool, psName string, contentType string,
+) {
+	childMessage := runtimePubsub.BulkSubscribeMessageItem{
+		Event:       event,
+		Metadata:    message.Metadata,
+		EntryID:     message.EntryID,
+		ContentType: contentType,
+	}
+	var cloudEvent map[string]interface{}
+	mapTypeEvent, ok := event.(map[string]interface{})
+	if ok {
+		cloudEvent = mapTypeEvent
+	}
+	if val, ok := (*routePathBulkMessageMap)[rPath]; ok {
+		if isCloudEvent {
+			val.cloudEvents[val.length] = mapTypeEvent
+		}
+		val.rawData[val.length] = childMessage
+		val.entries[val.length] = &msg.Entries[i]
+		val.length++
+		(*routePathBulkMessageMap)[rPath] = val
+	} else {
+		cloudEvents := make([]map[string]interface{}, len(msg.Entries))
+		rawDataItems := make([]runtimePubsub.BulkSubscribeMessageItem, len(msg.Entries))
+		rawDataItems[0] = childMessage
+		entries := make([]*pubsub.BulkMessageEntry, len(msg.Entries))
+		entries[0] = &msg.Entries[i]
+		if isCloudEvent {
+			cloudEvents[0] = cloudEvent
+		}
+		psm := pubsubBulkSubscribedMessage{
+			cloudEvents: cloudEvents,
+			rawData:     rawDataItems,
+			entries:     entries,
+			topic:       msg.Topic,
+			metadata:    msg.Metadata,
+			pubsub:      psName,
+			length:      1,
+		}
+		(*routePathBulkMessageMap)[rPath] = psm
+	}
+}
+
+func populateBulkSubscribeResponsesWithError(entries []*pubsub.BulkMessageEntry,
+	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, entryIDIndexMap *map[string]int, err error,
+) {
+	for _, item := range entries {
+		ind := (*entryIDIndexMap)[item.EntryID]
+		if (*bulkResponses)[ind].EntryID == "" {
+			(*bulkResponses)[ind].EntryID = item.EntryID
+			(*bulkResponses)[ind].Error = err
+		}
+	}
+}
+
+func populateAllBulkResponsesWithError(bulkMsg *pubsub.BulkMessage,
+	bulkResponses *[]pubsub.BulkSubscribeResponseEntry, err error,
+) {
+	for i, item := range bulkMsg.Entries {
+		if (*bulkResponses)[i].EntryID == "" {
+			(*bulkResponses)[i].EntryID = item.EntryID
+			(*bulkResponses)[i].Error = err
+		}
+	}
+}
+
+func newBulkSubIngressDiagnostics() bulkSubIngressDiagnostics {
+	statusWiseCountDiag := make(map[string]int64, 3)
+	statusWiseCountDiag[string(pubsub.Success)] = 0
+	statusWiseCountDiag[string(pubsub.Drop)] = 0
+	statusWiseCountDiag[string(pubsub.Retry)] = 0
+	bulkSubDiag := bulkSubIngressDiagnostics{
+		statusWiseDiag: statusWiseCountDiag,
+		elapsed:        0,
+		retryReported:  false,
+	}
+	return bulkSubDiag
+}
+
+func reportBulkSubDiagnostics(ctx context.Context, topic string, bulkSubDiag *bulkSubIngressDiagnostics) {
```

Comment:
Reviewers: This method is used to report back corresponding diagnostics. Sends number of events per status as well.